pkg/pool-manager/virtualmachine_pool.go

```diff
@@ -56,24 +56,32 @@ func (p *PoolManager) AllocateVirtualMachineMac(virtualMachine *kubevirt.Virtual
 		"namespace", virtualMachine.Namespace,
 		"interfaces", virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces)
 
-	for idx, iface := range virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces {
+	copyVM := virtualMachine.DeepCopy()
+	allocations := []string{}
+	for idx, iface := range copyVM.Spec.Template.Spec.Domain.Devices.Interfaces {
 		if iface.Masquerade == nil && iface.Slirp == nil && networks[iface.Name].Multus == nil {
 			log.Info("mac address can be set only for interface of type masquerade and slirp on the pod network")
 			continue
 		}
 
 		if iface.MacAddress != "" {
-			if err := p.allocateRequestedVirtualMachineInterfaceMac(iface.MacAddress, virtualMachine); err != nil {
+			if err := p.allocateRequestedVirtualMachineInterfaceMac(iface.MacAddress, copyVM); err != nil {
+				p.revertAllocationOnVm(vmNamespaced(copyVM), allocations)
 				return err
 			}
+			allocations = append(allocations, iface.MacAddress)
 		} else {
-			macAddr, err := p.allocateFromPoolForVirtualMachine(virtualMachine)
+			macAddr, err := p.allocateFromPoolForVirtualMachine(copyVM)
 			if err != nil {
+				p.revertAllocationOnVm(vmNamespaced(copyVM), allocations)
 				return err
 			}
-			virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces[idx].MacAddress = macAddr
+			copyVM.Spec.Template.Spec.Domain.Devices.Interfaces[idx].MacAddress = macAddr
+			allocations = append(allocations, macAddr)
 		}
 	}
+
+	virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces = copyVM.Spec.Template.Spec.Domain.Devices.Interfaces
```

Comment:
I'm little tired, could you please explain to me why do we need the copy?

Reply:
because we add the mac addresses allocated into the copy object so if we need to revert it will be easy (we don't change the original object)

if we get to this line the allocation succeeded so we copy the interfaces from the copy object (the one with the macAddress field configured and validated by the manager) to the original one.

Reply:
why don't we do the same with pods?

Reply:
in pods we marshal the networks again into the annotation :)